src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -69,259 +85,125 @@ std::string GetFullTypeWithSubTypes(const char * type, DnssdServiceProtocol prot
+        mData = TXTRecordGetBytesPtr(&mRecordRef);
+        return CHIP_NO_ERROR;
     }
 
-    return CHIP_NO_ERROR;
-}
+    uint16_t size() { return mDataSize; }
+    const void * data() { return mData; }
 
-bool CheckForSuccess(GenericContext * context, const char * name, DNSServiceErrorType err, bool useCallback = false)
-{
-    if (context == nullptr)
-    {
-        ChipLogError(DeviceLayer, "%s (%s)", name, "Mdns context is null.");
-        return false;
-    }
+private:
+    uint16_t mDataSize = 0;
+    const void * mData = nullptr;
 
-    if (kDNSServiceErr_NoError != err)
-    {
-        ChipLogError(DeviceLayer, "%s (%s)", name, Error::ToString(err));
+    TXTRecordRef mRecordRef;
+    char mRecordBuffer[kDnssdTextMaxSize];
+};
 
-        if (useCallback)
-        {
-            switch (context->type)
-            {
-            case ContextType::Register: {
-                RegisterContext * registerContext = reinterpret_cast<RegisterContext *>(context);
-                registerContext->callback(registerContext->context, nullptr, CHIP_ERROR_INTERNAL);
-                break;
-            }
-            case ContextType::Browse: {
-                BrowseContext * browseContext = reinterpret_cast<BrowseContext *>(context);
-                browseContext->callback(browseContext->context, nullptr, 0, CHIP_ERROR_INTERNAL);
-                break;
-            }
-            case ContextType::Resolve: {
-                ResolveContext * resolveContext = reinterpret_cast<ResolveContext *>(context);
-                resolveContext->callback(resolveContext->context, nullptr, Span<Inet::IPAddress>(), CHIP_ERROR_INTERNAL);
-                break;
-            }
-            case ContextType::GetAddrInfo: {
-                GetAddrInfoContext * resolveContext = reinterpret_cast<GetAddrInfoContext *>(context);
-                resolveContext->callback(resolveContext->context, nullptr, Span<Inet::IPAddress>(), CHIP_ERROR_INTERNAL);
-                break;
-            }
-            }
-        }
+} // namespace
 
-        if (CHIP_ERROR_KEY_NOT_FOUND == MdnsContexts::GetInstance().Remove(context))
-        {
-            chip::Platform::Delete(context);
-        }
+namespace chip {
+namespace Dnssd {
 
-        return false;
-    }
+MdnsContexts MdnsContexts::sInstance;
 
-    return true;
-}
+namespace {
 
 static void OnRegister(DNSServiceRef sdRef, DNSServiceFlags flags, DNSServiceErrorType err, const char * name, const char * type,
                        const char * domain, void * context)
 {
-    RegisterContext * sdCtx = reinterpret_cast<RegisterContext *>(context);
-    VerifyOrReturn(CheckForSuccess(sdCtx, __func__, err));
-
     ChipLogDetail(DeviceLayer, "Mdns: %s name: %s, type: %s, domain: %s, flags: %d", __func__, name, type, domain, flags);
 
-    sdCtx->callback(sdCtx->context, type, CHIP_NO_ERROR);
+    auto sdCtx = reinterpret_cast<RegisterContext *>(context);
+    sdCtx->Finalize(err);
```

Comment:
So just trying to trace the lifetimes:
1. If `sdCtx` is not in `MdnsContexts::GetInstance` this will delete it.
2. If it _is_ in `MdnsContexts` and this is a failure, it will call the callback and remove it, which calls `MdnsContexts::Delete` and hence deletes the object.
3. If `sdCtx` is in `MdnsContexts` and this is a success, we will call the callback and do nothing else; no removing.  Is that expected?

Reply:
1. Correct
2. Correct
3. This is correct for the "Register" type but for "Resolve" and "Browse" the callback also remove it. This is because in the case of "Register" we want to keep the context alive until "ChipDnssdRemoveServices" is called.